File: src/dpl_pipeline/segment.py
"""Split extracted bill text into provisions.

A "provision" here is a section-level unit: rules are evaluated per
provision so that (a) evidence excerpts point at the specific section that
triggered a category, and (b) proximity windows in rule regexes cannot
accidentally span unrelated sections.

Strategy:
1. Split on statutory section markers ("SECTION 1.", "Sec. 3.", "§ 24.2-643",
   "Be it enacted").
2. If no markers are found (common for flat PDF extractions), fall back to
   fixed-size paragraph chunks so long texts are still windowed.
"""

from __future__ import annotations

import re

from .models import Provision

# Line-anchored section markers.
_SECTION_RE = re.compile(
    r"^(?:"
    r"(?:SECTION|Section|Sec\.)\s+\d+[A-Za-z]?\s*[.:]"   # SECTION 1. / Sec. 2:
    r"|§+\s*[\d.]+[-\d.A-Za-z:]*"                    # § 24.2-643
    r"|Be it enacted\b"
    r")",
    re.MULTILINE,
)

_FALLBACK_CHUNK_CHARS = 2500
_MIN_PROVISION_CHARS = 40


def segment(text: str) -> list[Provision]:
    starts = [m.start() for m in _SECTION_RE.finditer(text)]
    if len(starts) >= 2:
        bounds = starts + [len(text)]
        chunks = []
        if starts[0] > 0:
            chunks.append(text[: starts[0]])  # preamble / title
        chunks.extend(text[bounds[i]: bounds[i + 1]] for i in range(len(starts)))
    else:
        chunks = _fallback_chunks(text)

    provisions: list[Provision] = []
    for chunk in chunks:
        chunk = chunk.strip()
        if len(chunk) < _MIN_PROVISION_CHARS:
            continue
        heading = chunk.split("\n", 1)[0][:120]
        provisions.append(
            Provision(
                provision_id=f"P{len(provisions) + 1:03d}",
                heading=heading,
                text=chunk,
            )
        )
    if not provisions and text.strip():
        provisions.append(Provision("P001", text.strip().split("\n", 1)[0][:120], text.strip()))
    return provisions


def _fallback_chunks(text: str) -> list[str]:
    paragraphs = re.split(r"\n\s*\n", text)
    chunks: list[str] = []
    current = ""
    for para in paragraphs:
        if current and len(current) + len(para) > _FALLBACK_CHUNK_CHARS:
            chunks.append(current)
            current = para
        else:
            current = f"{current}\n\n{para}" if current else para
    if current:
        chunks.append(current)
    return chunks

File: lib/segment.ts
/**
 * Split extracted bill text into provisions.
 *
 * A "provision" is a section-level unit: rules are evaluated per provision so
 * that (a) evidence excerpts point at the specific section that triggered a
 * category and (b) proximity windows in rule regexes cannot span unrelated
 * sections.
 *
 * Strategy: split on statutory section markers ("SECTION 1.", "Sec. 3.",
 * "§ 24.2-643", "Be it enacted"); if fewer than two markers are found (common
 * for flat PDF extractions), fall back to fixed-size paragraph chunks.
 */

import type { Provision } from "./types";

const SECTION_RE =
  /^(?:(?:SECTION|Section|Sec\.)\s+\d+[A-Za-z]?\s*[.:]|§+\s*[\d.][-\d.A-Za-z:]*|Be it enacted\b)/gm;

const FALLBACK_CHUNK_CHARS = 2500;
const MIN_PROVISION_CHARS = 40;

export function segment(text: string): Provision[] {
  const starts = [...text.matchAll(SECTION_RE)].map((m) => m.index);
  let chunks: string[];
  if (starts.length >= 2) {
    chunks = [];
    if (starts[0] > 0) chunks.push(text.slice(0, starts[0])); // preamble / title
    const bounds = [...starts, text.length];
    for (let i = 0; i < starts.length; i++) chunks.push(text.slice(bounds[i], bounds[i + 1]));
  } else {
    chunks = fallbackChunks(text);
  }

  const provisions: Provision[] = [];
  for (const rawChunk of chunks) {
    const chunk = rawChunk.trim();
    if (chunk.length < MIN_PROVISION_CHARS) continue;
    provisions.push({
      provisionId: `P${String(provisions.length + 1).padStart(3, "0")}`,
      heading: chunk.split("\n", 1)[0].slice(0, 120),
      text: chunk,
    });
  }
  if (provisions.length === 0 && text.trim()) {
    const t = text.trim();
    provisions.push({ provisionId: "P001", heading: t.split("\n", 1)[0].slice(0, 120), text: t });
  }
  return provisions;
}

function fallbackChunks(text: string): string[] {
  const paragraphs = text.split(/\n\s*\n/);
  const chunks: string[] = [];
  let current = "";
  for (const para of paragraphs) {
    if (current && current.length + para.length > FALLBACK_CHUNK_CHARS) {
      chunks.push(current);
      current = para;
    } else {
      current = current ? `${current}\n\n${para}` : para;
    }
  }
  if (current) chunks.push(current);
  return chunks;
}
